src/hierarchical.rs

```diff
@@ -0,0 +1,493 @@
+use std::cmp::{Ordering, min, max};
+use std::f64::INFINITY;
+use std::mem::replace;
+
+use {Point};
+
+
+/// A function to compute the inter-cluster distance between set `A` and `B`.
+pub type LinkageFunction = Fn(&Fn(usize, usize) -> f64, &[usize], &[usize]) -> f64;
+
+
+/// Hierarchical linkage criteria.
+///
+/// The linkage decides how the distance between clusters is computed.
+#[derive(Copy, Clone)]
+pub enum LinkageCriterion<'a> {
```

Comment:
I wonder if this enum may be better handled as a trait, to avoid the virtual dispatch of `Custom` and especially of the `&Fn(usize, usize)` (which, being a simple distance calculation, could take a few nanoseconds to compute, so the overhead of the virtual call can be significant).

The trait might look like:

```rust
trait LinkageCriterion {
    fn compute_dendrogram<'a, P: Point>(self, points: &[&'a P]) -> Box<Dendrogram<&'a P>> {
        naive(points, self)
    }

    fn linkage_function<F: Fn(usize, usize) -> f64>(&self, dist: F, a: &[usize], b: &[usize]) -> f64;
}
```

With `naive` adjusted to have signature something like 
```rust
fn naive<'a, P: Point, L: LinkageCriterion>(points: &[&'a P], linkage: L) ...
```
and call

```rust
linkage.linkage_function(|i,j| point_distances[max(i,j)][min(i,j)], &c.elements, &i.elements)
```

One would then define `pub struct Single;`, `pub struct CLINK;` (which override `compute_dendrogram` with `slink`/`clink`) and `pub struct Complete;` (which just customises `linkage_function`).

One downside of this approach is it forces code to be generic (since `LinkageCriterion` has generic methods it can't become a (useful) trait object), but pretty much all code here and in `cogset` as a whole is already generic over `P: Point` anyway, so it probably doesn't matter.